### PR TITLE
Correctly detach header/footer components when column is removed

### LINKED ISSUE #8638: Grid does not clean up components when columns removed
referencing [bug #8140](https://github.com/vaadin/framework/issues/8140)

Vaadin version: 7.7.7
when using a PropertySetChangeNotifier as DataSource, columns are removed/added automatically, and header cells are removed but not cleaned up.

example:

	package test.failure;
	
	import com.vaadin.data.Container.Indexed;
	import com.vaadin.data.util.IndexedContainer;
	import com.vaadin.ui.Button;
	import com.vaadin.ui.Grid;
	
	public class FailureGrid extends Grid {
	
		public FailureGrid() {
			Indexed i = new IndexedContainer();
			setContainerDataSource(i);
			i.addContainerProperty("c1", String.class, "does not matter 1");
			i.addContainerProperty("c3", String.class, "does not matter 2");
	
			HeaderRow h1 = prependHeaderRow();
			h1.getCell("c3").setComponent(new Button("This one will be an orthan"));
			h1.getCell("c1").setComponent(new Button("Click me to fail!", _ignore -> i.removeContainerProperty("c3")));
		}
	}

- just instantiate it, add to UI and press the first (left most) button in the header

hint:

	com.vaadin.ui.Grid.StaticSection.removeColumn(Object)

* cell is removed but not detached

### PR DESCRIPTION
**Pull request by Artur-**
Fixes #8638

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/8645)

### CODE REVIEW

File: server/src/test/java/com/vaadin/tests/server/component/grid/GridChildrenTest.java [hunk 1]
```diff
@@ -74,6 +79,93 @@ public void removeHeaderWithComponentInMergedHeaderCell() {
     }
 
     @Test
+    public void removeHeaderFooterComponentWhenColumnIsRemoved() {
+        HeaderRow h1 = grid.prependHeaderRow();
+        FooterRow f1 = grid.addFooterRowAt(0);
+        Button headerButton = new Button("This one will be an orthan");
```
Comment (tsuoanttila) [inline server/src/test/java/com/vaadin/tests/server/component/grid/GridChildrenTest.java]: Typo in button caption ?

**Comment by pleku** [comment]
Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8645)*

**Comment by pleku** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8645#-Ke7kMcqFzMKHlHnBjbF:-Ke7kMcqFzMKHlHnBjbG:bkox103) (no related file):*
:lgtm: but running modern browser tests http://r2d2.devnet.vaadin.com/viewQueued.html?itemId=151596&tab=queuedBuildOverviewTab

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8645)*

**Comment by Artur-** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8645#-Ke7kMcqFzMKHlHnBjbF:-KeTDNGRg1xcVpFD4lMu:b-vzvsxv) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

:lgtm: but running modern browser tests http://r2d2.devnet.vaadin.com/viewQueued.html?itemId=151596&tab=queuedBuildOverviewTab
</blockquote></details>

That build says "22 test failed (no new)"

---

*[server/src/test/java/com/vaadin/tests/server/component/grid/GridChildrenTest.java, line 85 at r1](https://reviewable.io:443/reviews/vaadin/framework/8645#-KdXExks-kJBWcJ_YSte-r1-85:-KeTDP5esi9wIfHkCXua:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/18cbc90e7b056d096efc505a43a0b3f1cbe0dc2c/server/src/test/java/com/vaadin/tests/server/component/grid/GridChildrenTest.java#L85)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Typo in button caption ?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8645)*

**Comment by pleku** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8645)*

**Pull request merged**